File: worker/providers/comfyui.py
from __future__ import annotations

import os
import time
from pathlib import Path
from typing import Any

import httpx


STYLE_PROMPTS: dict[str, dict[str, object]] = {
    'vintage-stop-motion-neon-2d': {
        'positive': 'vintage stop-motion animation, tactile puppetry, handcrafted miniatures, neon 2D linework, graphic motion design, cinematic mixed media',
        'negative': 'photorealistic, flat lighting, blurry, low quality, static frame, inconsistent motion',
        'denoise': 0.76,
    },
    'live-action-comic-halftone': {
        'positive': 'live-action footage transformed into comic book art, bold ink outlines, halftone shading, pop-art contrast, graphic novel styling',
        'negative': 'photorealistic, washed out, blurry, low contrast, soft shading',
        'denoise': 0.72,
    },
    'watercolor-cutout': {
        'positive': 'watercolor washes, cut-out collage shapes, layered paper texture, soft painterly motion, handcrafted mixed media',
        'negative': 'photorealistic, hard edges, digital noise, muddy colors, low detail',
        'denoise': 0.78,
    },
    'paper-collage-ink': {
        'positive': 'paper collage, torn paper edges, digital ink outlines, layered magazine texture, expressive mixed media animation',
        'negative': 'photorealistic, smooth surfaces, blurry, flat composition, low contrast',
        'denoise': 0.74,
    },
    'scrapbook-cutout-motion': {
        'positive': 'cohesive mixed media animation with all media layers clearly visible, real-world photography base blended with bold 2D cel animation linework, paper cutout collage layers, visible tactile 3D clay texture patches, woven fabric grain patches, corrugated cardboard edge texture, hand-drawn black ink marks, marker doodles, scribbled arrows and circles, visible sticker labels, torn masking tape strips attached as physical collage elements, scanned paper grain, rough scissor edges, imperfect white borders, subtle drop shadows between media layers, stop-motion jitter, frame-by-frame handmade animation energy, dynamic editorial composition, preserve source subject identity pose camera angle and readable silhouette',
        'negative': 'single medium only, pure photorealism, pure anime, pure 3D render, pure watercolor, pure oil painting, smooth CGI with no handmade texture, seamless digital compositing, no tape, no stickers, no doodles, no fabric texture, no cardboard texture, flat vector UI, warped anatomy, excessive typography over subject',
        'denoise': 0.79,
    },
    'glossy-3d-rotoscope': {
        'positive': 'glossy 3D surfaces, rotoscoped motion, cinematic lighting, stylized depth, polished mixed media animation',
        'negative': 'low poly, flat shading, noisy texture, blurry motion, inconsistent geometry',
        'denoise': 0.68,
    },
    'vhs-neon-sketch': {
        'positive': 'VHS glitch, neon sketch lines, retro analog distortion, chromatic aberration, synthwave mixed media animation',
        'negative': 'clean digital video, photorealistic, soft blur, modern pristine look, low contrast',
        'denoise': 0.75,
    },
}

STYLE_FALLBACKS: list[tuple[tuple[str, ...], str]] = [
    (('scrapbook', 'mixed media', 'collage', 'cutout', 'paper', 'tape', 'sticker', 'doodle', 'layer'), 'scrapbook-cutout-motion'),
    (('comic', 'halftone', 'graphic', 'poster', 'motion design', 'this is fine', 'graffiti', 'paparazzi', 'grime', 'two color'), 'live-action-comic-halftone'),
    (('watercolor', 'ghibli', 'storybook', 'cloud', 'dreamy', 'canvas', 'oil painting'), 'watercolor-cutout'),
    (('anime', 'claymation', 'gouache', 'sketch', 'ink', 'risograph'), 'paper-collage-ink'),
    (('vhs', 'retro', 'pixel', '60s', 'japanese', 'burning sunset'), 'vhs-neon-sketch'),
    (('noir', 'giallo', 'neon', 'cyberpunk', 'dragon', 'night vision', 'office', 'red carpet', 'particles', 'multiverse', 'glitch'), 'glossy-3d-rotoscope'),
]


class ComfyUIClient:
    def __init__(self, base_url: str | None = None) -> None:
        self.base_url = (base_url or os.getenv('COMFYUI_BASE_URL') or 'http://127.0.0.1:8188').rstrip('/')
        self.checkpoint = os.getenv('COMFYUI_CHECKPOINT', 'sdxl.safetensors')
        self.api_key = os.getenv('COMFYUI_API_KEY') or os.getenv('AI_STYLE_API_KEY') or ''
        self.auth_header = os.getenv('COMFYUI_API_KEY_HEADER', 'Authorization')
        default_prefix = 'Bearer' if self.auth_header.lower() == 'authorization' else ''
        self.auth_prefix = os.getenv('COMFYUI_API_KEY_PREFIX', default_prefix)
        self.client = httpx.Client(timeout=300.0, headers=self._auth_headers())

    def _auth_headers(self) -> dict[str, str]:
        if not self.api_key:
            return {}
        value = f'{self.auth_prefix} {self.api_key}'.strip() if self.auth_prefix else self.api_key
        return {self.auth_header: value}

    def _resolve_style(self, style: str, prompt: str = '') -> dict[str, object]:
        normalized = style.strip().lower()
        if normalized in STYLE_PROMPTS:
            return STYLE_PROMPTS[normalized]

        search_text = f'{style} {prompt}'.lower()
        for keywords, fallback_style in STYLE_FALLBACKS:
            if any(keyword in search_text for keyword in keywords):
                return STYLE_PROMPTS[fallback_style]

        return STYLE_PROMPTS['scrapbook-cutout-motion']

    def upload_image(self, image_path: str) -> str:
        with open(image_path, 'rb') as file_handle:
            response = self.client.post(
                f'{self.base_url}/upload/image',
                files={'image': (Path(image_path).name, file_handle, 'image/png')},
            )
        response.raise_for_status()
        payload = response.json()
        return payload['name']

    def _default_workflow(self, uploaded_image_name: str, style: str, prompt: str, strength: float | None = None) -> dict[str, Any]:
        style_config = self._resolve_style(style, prompt)
        style_prompt = str(style_config['positive'])
        extra_prompt = prompt.strip()
        positive_prompt = f'{style_prompt}, {extra_prompt}' if extra_prompt else style_prompt
        negative_prompt = str(style_config['negative'])
        denoise = max(0.0, min(1.0, strength if strength is not None else float(style_config['denoise'])))
        return {
            '1': {
                'class_type': 'LoadImage',
                'inputs': {'image': uploaded_image_name},
            },
            '2': {
                'class_type': 'CheckpointLoaderSimple',
                'inputs': {'ckpt_name': self.checkpoint},
            },
            '3': {
                'class_type': 'CLIPTextEncode',
                'inputs': {'text': positive_prompt, 'clip': ['2', 1]},
            },
            '4': {
                'class_type': 'CLIPTextEncode',
                'inputs': {'text': negative_prompt, 'clip': ['2', 1]},
            },
            '5': {
                'class_type': 'VAEEncode',
                'inputs': {'pixels': ['1', 0], 'vae': ['2', 2]},
            },
            '6': {
                'class_type': 'KSampler',
                'inputs': {
                    'seed': 123456789,
                    'steps': 20,
                    'cfg': 7,
                    'sampler_name': 'dpmpp_2m',
                    'scheduler': 'karras',
                    'denoise': denoise,
                    'model': ['2', 0],
                    'positive': ['3', 0],
                    'negative': ['4', 0],
                    'latent_image': ['5', 0],
                },
            },
            '7': {
                'class_type': 'VAEDecode',
                'inputs': {'samples': ['6', 0], 'vae': ['2', 2]},
            },
            '8': {
                'class_type': 'SaveImage',
                'inputs': {'filename_prefix': 'morphix/frame', 'images': ['7', 0]},
            },
        }

    def submit_frame(self, image_path: str, style: str, prompt: str = '', strength: float | None = None) -> str:
        image_name = self.upload_image(image_path)
        workflow = self._default_workflow(image_name, style, prompt, strength)
        response = self.client.post(
            f'{self.base_url}/prompt',
            json={'prompt': workflow, 'client_id': os.getenv('COMFYUI_CLIENT_ID', 'morphix-worker')},
        )
        response.raise_for_status()
        return response.json()['prompt_id']

    def wait_for_output(self, prompt_id: str) -> list[dict[str, Any]]:
        deadline = time.time() + 1800
        while time.time() < deadline:
            response = self.client.get(f'{self.base_url}/history/{prompt_id}')
            response.raise_for_status()
            payload = response.json()
            if prompt_id in payload:
                outputs: list[dict[str, Any]] = []
                nodes = payload[prompt_id].get('outputs', {})
                for node in nodes.values():
                    outputs.extend(node.get('images', []))
                if outputs:
                    return outputs
            time.sleep(2)
        raise TimeoutError(f'ComfyUI job {prompt_id} timed out')

    def download_output(self, image_info: dict[str, Any], target_path: str) -> None:
        response = self.client.get(
            f"{self.base_url}/view",
            params={
                'filename': image_info['filename'],
                'subfolder': image_info.get('subfolder', ''),
                'type': image_info.get('type', 'output'),
            },
        )
        response.raise_for_status()
        Path(target_path).parent.mkdir(parents=True, exist_ok=True)
        with open(target_path, 'wb') as file_handle:
            file_handle.write(response.content)

File: lib/server/comfyui.ts
import fs from 'fs';
import path from 'path';

const BASE_URL = (process.env.COMFYUI_BASE_URL ?? 'http://127.0.0.1:8188').replace(/\/$/, '');
const CHECKPOINT = process.env.COMFYUI_CHECKPOINT ?? 'sdxl.safetensors';
const CLIENT_ID = process.env.COMFYUI_CLIENT_ID ?? 'morphix-next';

interface StylePrompt {
  positive: string;
  negative: string;
  denoise: number;
  steps?: number;
  cfg?: number;
  sampler?: string;
  scheduler?: string;
}

interface ResultImageInfo {
  filename: string;
  subfolder?: string;
  type?: string;
}

const COMMON_POSITIVE =
  'single frame from a mixed-media animation, preserve source composition, pose, character identity, camera angle, clean silhouettes, temporally consistent details';

const COMMON_NEGATIVE =
  'raw unstyled video frame, flicker, inconsistent character design, warped anatomy, duplicate limbs, blurry, low quality, text, watermark, logo, frame border';

export const STYLE_PROMPTS: Record<string, StylePrompt> = {
  // ── Artistic ─────────────────────────────────────────────────
  anime: {
    positive: 'anime animation, bold cel shading, vibrant color keys, clean sharp line art, expressive shapes, hand-painted background',
    negative: 'photorealistic, 3d render, blurry, washed out, muddy lines',
    denoise: 0.72, cfg: 7.2,
  },
  ghibli: {
    positive: 'studio ghibli style, warm hand-drawn fantasy, lush painted nature backgrounds, magical realism, soft colors, gentle expressive characters',
    negative: 'dark gritty photorealism, 3d render, harsh neon, sterile studio photo, sharp digital',
    denoise: 0.70, cfg: 7.0,
  },
  watercolor: {
    positive: 'watercolor animation, soft washes, wet-on-wet pigment blooms, translucent layers, delicate paper texture, pastel rhythm',
    negative: 'sharp plastic edges, photorealistic, harsh digital render, oversaturated, flat vector',
    denoise: 0.70, cfg: 7.0,
  },
  'oil-painting': {
    positive: 'oil painting, thick impasto brushwork, rich chiaroscuro, museum quality, visible canvas texture, glazed depth',
    negative: 'flat digital, watercolor, thin lines, photorealistic photo, low quality',
    denoise: 0.74, cfg: 7.4,
  },
  sketch: {
    positive: 'hand-drawn pencil sketch, black and white line art, detailed cross-hatching, expressive construction lines, animated storyboard',
    negative: 'photorealistic, saturated color, glossy digital painting, filled color blocks',
    denoise: 0.68, cfg: 7.0,
  },
  ink: {
    positive: 'sumi-e ink brush, black ink wash, East Asian calligraphy style, expressive brushwork, minimal negative space',
    negative: 'color photography, flat digital, watercolor wash, pastel, low contrast',
    denoise: 0.66, cfg: 7.5,
  },
  gouache: {
    positive: 'gouache illustration, opaque flat matte color, editorial poster art, crisp edges, designer palette',
    negative: 'transparent watercolor, photorealistic, blurry, gradient noise, 3d render',
    denoise: 0.68, cfg: 7.2,
  },
  risograph: {
    positive: 'risograph print, halftone dots, misregistration, duotone ink layers, tactile paper, limited palette',
    negative: 'seamless digital, photorealistic, full-color photo, glossy, smooth',
    denoise: 0.70, cfg: 7.4,
  },
  claymation: {
    positive: 'claymation stop-motion, clay material, plasticine fingerprints, chunky handcrafted surfaces, warm miniature set lighting',
    negative: 'photorealistic, sharp digital render, flat 2d only, CG animation',
    denoise: 0.74, cfg: 7.2,
  },
  collage: {
    positive: 'mixed-media collage, torn magazine paper, layered ephemera, pop-art composition, tactile handmade depth',
    negative: 'photorealistic, seamless, uniform background, plain monotone',
    denoise: 0.76, cfg: 7.5,
  },
  'scrapbook-cutout-motion': {
    positive: [
      'cohesive mixed media animation',
      'all requested media layers must be visibly present in the frame',
      'real-world photography base transformed into a hybrid animated frame',
      '2D cel animation linework over live-action forms, clean graphic outlines and selective color fills',
      'paper cutout collage layers, torn poster paper, magazine fragments, tape, stickers, and zine textures',
      'visible tactile 3D material accents, clay texture patches, woven fabric grain, corrugated cardboard edges, small sculptural props',
      'hand-drawn ink marks, marker doodles, scribble accents, arrows, circles, and rough underline marks',
      'stickers and torn masking tape must appear as physical collage elements attached to the composition',
      'scanned paper grain, glue shadows, imperfect white borders, rough scissor edges',
      'layered depth with subtle drop shadows between photography, cut paper, cel lines, and 3D texture',
      'stop-motion jitter and frame-by-frame handmade animation energy',
      'dynamic editorial composition, visually unified mixed-media prompt-generator style',
      'preserve source subject identity, pose, camera angle, and readable silhouette',
    ].join(', '),
    negative: [
      'single medium only, pure photorealism, pure anime, pure 3D render, pure watercolor, pure oil painting',
      'smooth CGI with no handmade texture, seamless digital compositing, glossy plastic, no tape, no stickers, no doodles, no fabric texture, no cardboard texture',
      'flat vector UI, minimal corporate presentation style, plain poster design',
      'vintage sepia-only Victorian collage, monochrome old newspaper only',
      'messy unreadable face, warped anatomy, excessive typography over subject',
    ].join(', '),
    denoise: 0.79,
    cfg: 7.8,
    steps: 28,
    sampler: 'dpmpp_2m',
    scheduler: 'karras',
  },
  canvas: {
    positive: 'acrylic on canvas, visible rough canvas weave, painterly brush texture, warm tones, museum quality',
    negative: 'flat digital, watercolor, thin sketch, plastic surfaces, photo',
    denoise: 0.72, cfg: 7.0,
  },
  storybook: {
    positive: 'vintage storybook illustration, pastel palette, nostalgic warmth, soft ink outlines, Caldecott style',
    negative: 'dark gritty, photorealistic, harsh contrast, neon, sharp digital',
    denoise: 0.68, cfg: 6.8,
  },

  'cutout-collage-animation': {
    positive: [
      'Terry Gilliam Monty Python animated collage style',
      'vintage magazine photograph cutouts arranged as flat layered paper planes',
      'Victorian anatomical diagrams mixed with 1960s pop culture imagery',
      'surrealist editorial collage with anachronistic mashup of historical and modern imagery',
      'bold hand-lettered title typography cut from newspaper headlines',
      'stop-motion jitter aesthetic with visible misalignment between paper elements',
      '16mm film grain overlay, off-white aged newsprint background texture',
      'flat poster-paint accent color fills, hard drop shadows cast between paper layers',
      'visible torn and scissor-cut paper edges, rough white borders around cutout figures',
      'mixed scale — tiny figures next to giant objects, impossible juxtapositions',
      'single frame from a mixed-media paper cutout animation',
      'temporally consistent paper plane positions, preserve character cutout identity',
    ].join(', '),
    negative: [
      'photorealistic 3D render, smooth CGI animation, seamless digital blend',
      'full digital painting, watercolor wash, oil paint texture, smooth brushwork',
      'clean modern minimal UI design, flat vector iconography',
      'dark cinematic atmosphere, volumetric fog, atmospheric perspective',
      'anime cel shading, cartoon outlines, comic book halftone',
      'color photography with no post-processing, realistic skin texture',
      'contemporary digital collage, clean Photoshop compositing',
    ].join(', '),
    denoise: 0.78,
    cfg: 7.8,
    steps: 28,
    sampler: 'dpmpp_2m',
    scheduler: 'karras',
  },

  // ── Cinematic ─────────────────────────────────────────────────
  noir: {
    positive: 'film noir, high contrast black and white, dramatic directional shadows, rain-slick streets, detective cinema mood, inked light shapes',
    negative: 'bright cheerful colors, soft pastel, low contrast, flat lighting, color',
    denoise: 0.64, cfg: 7.5,
  },
  giallo: {
    positive: 'giallo horror, deep saturated blood-orange tones, dramatic lighting, 70s Italian cinema, expressionist color',
    negative: 'desaturated, flat lighting, modern clean digital, black and white',
    denoise: 0.70, cfg: 7.4,
  },
  'neon-city': {
    positive: 'neon city, wet pavement reflections, cyberpunk rain lighting, electric signage, dark urban scene',
    negative: 'natural daylight, pastoral, warm plain interior, dull colors, flat lighting',
    denoise: 0.72, cfg: 7.2,
  },
  cyberpunk: {
    positive: 'cyberpunk, neon-lit megacity, chrome and circuitry, dystopian future, electric magenta and cyan glow',
    negative: 'natural daylight, warm analog, hand-drawn, clean minimal, pastel',
    denoise: 0.72, cfg: 7.5,
  },
  'dragon-fantasy': {
    positive: 'epic fantasy, dramatic god rays, painterly cinematic, myth and legend, heroic scale, volumetric light',
    negative: 'mundane setting, flat lighting, pastel, kawaii, slice of life',
    denoise: 0.70, cfg: 7.2,
  },
  'night-vision': {
    positive: 'night vision goggles, thermal green phosphor, grainy surveillance footage, NVGS grain overlay',
    negative: 'color daylight, warm interior, clean sharp image, no grain',
    denoise: 0.62, cfg: 7.0,
  },
  'office-cctv': {
    positive: 'CCTV security footage, timestamp overlay, scan lines, washed color, low resolution surveillance',
    negative: 'high resolution, sharp crisp detail, warm color grading, cinematic',
    denoise: 0.60, cfg: 6.8,
  },
  'red-carpet': {
    positive: 'red carpet fashion editorial, paparazzi flash photography, luxury glamour, high contrast magazine',
    negative: 'dark moody, rough street, low quality, blurry, candid snapshot',
    denoise: 0.64, cfg: 7.0,
  },

  // ── Retro & Vintage ───────────────────────────────────────────
  vhs: {
    positive: 'VHS tape, chroma aberration, magnetic noise, 80s home video, scan line distortion, retro consumer footage',
    negative: 'sharp modern HD, sterile digital render, no grain, clean 4K',
    denoise: 0.70, cfg: 7.3,
  },
  '60s-cafe': {
    positive: '1960s retro illustration, warm mid-century palette, coffee shop poster, Saul Bass influence',
    negative: 'modern digital, cold color, flat vector only, sharp photo',
    denoise: 0.68, cfg: 7.0,
  },
  'japanese-show': {
    positive: '90s Japanese television aesthetic, bold typography, noisy color, variety show energy, saturated',
    negative: 'minimal, cool modern, dark noir, black and white, desaturated',
    denoise: 0.68, cfg: 7.2,
  },
  pixel: {
    positive: 'pixel art, 16-bit sprites, limited color palette, dithering, retro game aesthetic',
    negative: 'smooth gradients, photorealistic, anti-aliased, modern 3d, blurry',
    denoise: 0.66, cfg: 7.5,
  },
  'two-color': {
    positive: 'duotone silkscreen, two color poster print, bold graphic, risograph-adjacent, punchy ink',
    negative: 'full color spectrum, subtle tones, photorealistic, smooth gradient',
    denoise: 0.72, cfg: 7.6,
  },
  'burnout-sunset': {
    positive: 'golden hour, lens flare, overexposed warm orange haze, dreamy sunset glow',
    negative: 'cold blue, night scene, studio lighting, flat neutral, gray',
    denoise: 0.64, cfg: 6.8,
  },

  // ── Street & Urban ─────────────────────────────────────────
  grime: {
    positive: 'grime street art, xerox texture, urban grainy, raw concrete, underground energy',
    negative: 'clean studio, smooth digital, soft pastel, refined luxury, nature',
    denoise: 0.70, cfg: 7.3,
  },
  graffiti: {
    positive: 'graffiti wildstyle, spray paint drips, tag letters, wall art, urban mural energy',
    negative: 'clean studio, watercolor, nature, subtle pastel, photorealistic',
    denoise: 0.74, cfg: 7.5,
  },
  neon: {
    positive: 'neon sign glow, glass tube electric hum, dark brick background, bent neon tubing',
    negative: 'natural daylight, warm analog, flat vector, no glow, subtle lighting',
    denoise: 0.70, cfg: 7.4,
  },
  paparazzi: {
    positive: 'candid paparazzi photo, flash photography, tabloid grain, street celebrity shot',
    negative: 'studio controlled, clean portrait, smooth skin, dark moody',
    denoise: 0.60, cfg: 6.8,
  },

  // ── Surreal & Abstract ─────────────────────────────────────
  particles: {
    positive: 'luminous particles, data visualization, constellation motion, light trails, kinetic energy',
    negative: 'dark flat, realistic photo, no glow, low energy, static',
    denoise: 0.72, cfg: 7.2,
  },
  multiverse: {
    positive: 'multiverse portal, dimensional rift, infinite recursion, surreal space, cosmic energy',
    negative: 'mundane everyday, realistic, flat lighting, single dimension, boring',
    denoise: 0.78, cfg: 7.6,
  },
  'layer-mixed': {
    positive: 'mixed media soul layers, photo collage with paint and text overlay, chaotic elegance, deep texture',
    negative: 'clean minimal, photorealistic portrait, simple flat, single medium',
    denoise: 0.76, cfg: 7.5,
  },
  glitch: {
    positive: 'digital glitch art, RGB channel split, data corruption, scan displacement, chromatic aberration',
    negative: 'clean smooth digital, photorealistic, no artifacts, stable image',
    denoise: 0.74, cfg: 7.4,
  },
  'doodle-melt': {
    positive: 'hand-drawn doodle melting into live action, cartoon physics, surreal animated blend',
    negative: 'photorealistic, fully abstract, no recognizable subject, clean digital',
    denoise: 0.72, cfg: 7.2,
  },
  'cloud-surf': {
    positive: 'volumetric clouds, pastel dreamscape, floaty surreal, heaven-like atmosphere, soft billowing',
    negative: 'dark storm, gritty urban, sharp hard edges, dark moody, noir',
    denoise: 0.68, cfg: 6.8,
  },

  // ── Graphic ────────────────────────────────────────────────
  comic: {
    positive: 'comic book panel, ben-day dots, bold ink outlines, action lines, speech bubble panel',
    negative: 'photorealistic, watercolor wash, painted texture, realistic skin, soft edges',
    denoise: 0.72, cfg: 7.5,
  },
  'sketch-to-real': {
    positive: 'pencil sketch morphing into photorealistic detail, drawing-to-real transition aesthetic',
    negative: 'fully digital, flat vector, no pencil marks, cartoon cel shaded',
    denoise: 0.68, cfg: 7.0,
  },
  sticker: {
    positive: 'kawaii sticker sheet, glitter, die-cut white outline, pastel, cute character design',
    negative: 'dark gritty, photorealistic, raw sketch, noir, no outline',
    denoise: 0.70, cfg: 7.0,
  },
  'this-is-fine': {
    positive: 'flat web comic meme art, simple outlines, internet humor aesthetic, editorial cartoon',
    negative: 'photorealistic, complex rendering, 3D, detailed texture',
    denoise: 0.68, cfg: 7.2,
  },
  'motion-design': {
    positive: 'motion design, geometric vector shapes, kinetic typography, clean animation, brand design',
    negative: 'photorealistic, rough texture, organic, painted, hand-drawn sketch',
    denoise: 0.64, cfg: 7.0,
  },

  // ── Legacy keys kept for backward compat ───────────────────
  'vintage-stop-motion-neon-2d': {
    positive: 'vintage stop-motion mixed-media animation, clay and paper surface texture, neon 2D cel outlines, handcrafted props',
    negative: 'photorealistic, glossy plastic, flat digital render, sterile studio lighting',
    denoise: 0.66, cfg: 7.2,
  },
  'live-action-comic-halftone': {
    positive: 'live action transformed into comic book halftone, bold ink contours, screen-printed dot shadows, pop art color blocking',
    negative: 'soft watercolor, plain photo, muddy colors, weak outlines',
    denoise: 0.62, cfg: 7.5,
  },
  'watercolor-cutout': {
    positive: 'watercolor wash mixed with cut-out paper collage, soft pigment blooms, layered paper edges, handmade shadows',
    negative: 'hard 3d render, neon cyberpunk, harsh ink, over-sharpened',
    denoise: 0.68, cfg: 7.0,
  },
  'paper-collage-ink': {
    positive: 'paper collage with digital ink, torn magazine textures, layered handmade paper, crisp black ink accents',
    negative: 'photorealistic, seamless airbrushed render, flat vector only',
    denoise: 0.70, cfg: 7.4,
  },
  'glossy-3d-rotoscope': {
    positive: 'glossy stylized 3D mixed with rotoscope, preserved body motion, sculpted depth, clean contour tracing',
    negative: 'flat sketch only, messy linework, clay texture, paper collage',
    denoise: 0.58, cfg: 6.8,
  },
  'vhs-neon-sketch': {
    positive: 'VHS neon sketch, analog tape texture, electric hand-drawn outlines, chromatic glow, scanline grain',
    negative: 'clean modern HD, plain photo, watercolor wash, muted palette',
    denoise: 0.64, cfg: 7.3,
  },
};

const STYLE_ALIASES: Record<string, string> = {
  'vintage stop motion neon 2d': 'vintage-stop-motion-neon-2d',
  'vintage stop-motion neon 2d': 'vintage-stop-motion-neon-2d',
  'stop motion neon': 'vintage-stop-motion-neon-2d',
  'live action comic halftone': 'live-action-comic-halftone',
  'comic halftone': 'live-action-comic-halftone',
  'watercolor cutout': 'watercolor-cutout',
  'watercolor cut-out': 'watercolor-cutout',
  'paper collage ink': 'paper-collage-ink',
  'scrapbook cutout motion': 'scrapbook-cutout-motion',
  'scrapbook cut-out motion': 'scrapbook-cutout-motion',
  'mixed media scrapbook': 'scrapbook-cutout-motion',
  'glossy 3d rotoscope': 'glossy-3d-rotoscope',
  'vhs neon sketch': 'vhs-neon-sketch',
};

function configuredApiKey() {
  return process.env.COMFYUI_API_KEY?.trim() || process.env.AI_STYLE_API_KEY?.trim() || '';
}

function authHeaders(extra: Record<string, string> = {}) {
  const apiKey = configuredApiKey();
  if (!apiKey) return extra;

  const headerName = process.env.COMFYUI_API_KEY_HEADER?.trim() || 'Authorization';
  const configuredPrefix = process.env.COMFYUI_API_KEY_PREFIX;
  const defaultPrefix = headerName.toLowerCase() === 'authorization' ? 'Bearer' : '';
  const prefix = configuredPrefix ?? defaultPrefix;
  const headerValue =
    prefix && !apiKey.toLowerCase().startsWith(`${prefix.toLowerCase()} `)
      ? `${prefix} ${apiKey}`
      : apiKey;

  return { ...extra, [headerName]: headerValue };
}

function normalizeStyleKey(style: string) {
  const key = style.trim().toLowerCase();
  if (STYLE_PROMPTS[key]) return key;
  if (STYLE_ALIASES[key]) return STYLE_ALIASES[key];
  const slug = key.replace(/[^a-z0-9]+/g, '-').replace(/^-+|-+$/g, '');
  return STYLE_PROMPTS[slug] ? slug : 'anime';
}

function resolveStylePrompt(style: string, promptOverride?: string, strength?: number) {
  const styleKey = normalizeStyleKey(style);
  const styleConfig = STYLE_PROMPTS[styleKey];
  const userPrompt = promptOverride?.trim();

  return {
    ...styleConfig,
    styleKey,
    // If caller passes strength, override the stored denoise value
    denoise: strength !== undefined ? Math.max(0, Math.min(1, strength)) : styleConfig.denoise,
    positive: [styleConfig.positive, COMMON_POSITIVE, userPrompt].filter(Boolean).join(', '),
    negative: [styleConfig.negative, COMMON_NEGATIVE].filter(Boolean).join(', '),
  };
}

export function getComfyUIConfig() {
  return {
    baseUrl: BASE_URL,
    checkpoint: CHECKPOINT,
    provider: process.env.MORPHIX_PROVIDER ?? process.env.AI_PROVIDER ?? 'mock',
    apiKeyConfigured: Boolean(configuredApiKey()),
    apiKeyHeader: configuredApiKey()
      ? process.env.COMFYUI_API_KEY_HEADER?.trim() || 'Authorization'
      : undefined,
  };
}

export async function pingComfyUI(): Promise<boolean> {
  try {
    const res = await fetch(`${BASE_URL}/system_stats`, {
      headers: authHeaders(),
      cache: 'no-store',
    });
    return res.ok;
  } catch {
    return false;
  }
}

function buildImg2ImgWorkflow(
  uploadedFilename: string,
  style: string,
  prompt?: string,
  strength?: number,
): Record<string, unknown> {
  const styleConfig = resolveStylePrompt(style, prompt, strength);

  return {
    '1': { class_type: 'CheckpointLoaderSimple', inputs: { ckpt_name: CHECKPOINT } },
    '2': { class_type: 'CLIPTextEncode', inputs: { text: styleConfig.positive, clip: ['1', 1] } },
    '3': { class_type: 'CLIPTextEncode', inputs: { text: styleConfig.negative, clip: ['1', 1] } },
    '4': { class_type: 'LoadImage', inputs: { image: uploadedFilename } },
    '5': { class_type: 'VAEEncode', inputs: { pixels: ['4', 0], vae: ['1', 2] } },
    '6': {
      class_type: 'KSampler',
      inputs: {
        seed: Math.floor(Math.random() * 1e12),
        steps: styleConfig.steps ?? Number(process.env.COMFYUI_STEPS ?? 22),
        cfg: styleConfig.cfg ?? Number(process.env.COMFYUI_CFG ?? 7),
        sampler_name: styleConfig.sampler ?? process.env.COMFYUI_SAMPLER ?? 'dpmpp_2m',
        scheduler: styleConfig.scheduler ?? process.env.COMFYUI_SCHEDULER ?? 'karras',
        denoise: styleConfig.denoise,
        model: ['1', 0],
        positive: ['2', 0],
        negative: ['3', 0],
        latent_image: ['5', 0],
      },
    },
    '7': { class_type: 'VAEDecode', inputs: { samples: ['6', 0], vae: ['1', 2] } },
    '8': {
      class_type: 'SaveImage',
      inputs: {
        filename_prefix: `morphix/${styleConfig.styleKey}`,
        images: ['7', 0],
      },
    },
  };
}

async function uploadFrame(framePath: string): Promise<string> {
  const form = new FormData();
  const imageBytes = new Uint8Array(await fs.promises.readFile(framePath));
  form.append('image', new Blob([imageBytes], { type: 'image/png' }), path.basename(framePath));

  const res = await fetch(`${BASE_URL}/upload/image`, {
    method: 'POST',
    headers: authHeaders(),
    body: form,
  });

  if (!res.ok) throw new Error(`ComfyUI upload failed: ${res.status} ${await res.text()}`);
  const json = (await res.json()) as { name: string; subfolder?: string };
  return json.subfolder ? `${json.subfolder}/${json.name}` : json.name;
}

async function queuePrompt(workflow: Record<string, unknown>): Promise<string> {
  const res = await fetch(`${BASE_URL}/prompt`, {
    method: 'POST',
    headers: authHeaders({ 'Content-Type': 'application/json' }),
    body: JSON.stringify({ prompt: workflow, client_id: CLIENT_ID }),
  });

  if (!res.ok) throw new Error(`ComfyUI queue failed: ${res.status} ${await res.text()}`);
  const json = (await res.json()) as { prompt_id: string };
  return json.prompt_id;
}

async function waitForResult(promptId: string, timeoutMs = 180_000): Promise<ResultImageInfo> {
  const deadline = Date.now() + timeoutMs;

  while (Date.now() < deadline) {
    await new Promise((resolve) => setTimeout(resolve, 1200));

    const res = await fetch(`${BASE_URL}/history/${promptId}`, {
      headers: authHeaders(),
      cache: 'no-store',
    });
    if (!res.ok) continue;

    const history = (await res.json()) as Record<
      string,
      {
        status?: { status_str?: string };
        outputs?: Record<string, { images?: ResultImageInfo[] }>;
      }
    >;

    const entry = history[promptId];
    if (!entry) continue;
    if (entry.status?.status_str === 'error') throw new Error('ComfyUI returned error status');

    const images = Object.values(entry.outputs ?? {}).flatMap(
      (nodeOutput) => nodeOutput.images ?? [],
    );
    const outputImage = images.find((image) => image.type === 'output') ?? images[0];
    if (outputImage) return outputImage;
  }

  throw new Error(`ComfyUI timed out after ${timeoutMs}ms`);
}

async function downloadResult(image: ResultImageInfo): Promise<Buffer> {
  const params = new URLSearchParams({
    filename: image.filename,
    type: image.type ?? 'output',
  });
  if (image.subfolder) params.set('subfolder', image.subfolder);

  const res = await fetch(`${BASE_URL}/view?${params.toString()}`, {
    headers: authHeaders(),
    cache: 'no-store',
  });

  if (!res.ok) throw new Error(`ComfyUI download failed: ${res.status} ${await res.text()}`);
  return Buffer.from(await res.arrayBuffer());
}

export async function stylizeFrame(
  framePath: string,
  style: string,
  prompt?: string,
  strength?: number,
): Promise<Buffer> {
  const uploadedFilename = await uploadFrame(framePath);
  const workflow = buildImg2ImgWorkflow(uploadedFilename, style, prompt, strength);
  const promptId = await queuePrompt(workflow);
  const image = await waitForResult(promptId);
  return downloadResult(image);
}

export default {
  getComfyUIConfig,
  pingComfyUI,
  stylizeFrame,
  STYLE_PROMPTS,
};
